[Apache Tomcat] Add integration package with memory data stream

harnish-elastic opened the pull request:
- Enhancement

## What does this PR do?

- Added 1 data stream (Memory metrics).
- Added data collection logic for the data streams.
- Added the ingest pipeline for the data streams.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yaml files.
- Added dashboards and visualizations.
- Added system test cases for the data stream.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] Verify if field mapping is correct in the data stream template.
- [x] Verification of data in visualisation after enabling TSDB flag in kibana
- [x] Verification of the count of documents (before & after TSDB enablement) in Discover Interface

## How to test this PR locally

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/apache_tomcat) directory.
- Run the following command to run tests. elastic-package test


## Screenshots
![apache_tomcat-memory-dashboard](https://github.com/elastic/integrations/assets/118714680/b7b4c20e-f10e-4807-b082-2be51d81f64f)


## Meta issue
- #5200

## Related PRs
- #5424
- #5926
- #5883
- #5983
- #6003
- #6573
- #6596
- #6609
<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6527/detail/PR-6527/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6527/detail/PR-6527/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6527/detail/PR-6527/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6527/detail/PR-6527/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_6527.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-06-09T17:52:52.029Z&rangeTo=2023-06-09T18:12:52.029Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=cdd75ee43997a0835c49e4c67665835f&transactionId=c61e6adbc5838c35)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-06-09T18:02:52.029+0000


* Duration: 21 min 35 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 36  |
| Skipped      | 0 |
| Total        | 36   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`6/6`) | :green_heart: 
Files | 100.0% (`6/6`) | :green_heart: 6.25
Classes | 100.0% (`6/6`) | :green_heart: 6.25
Methods | 93.22% (`55/59`) | :+1: 7.3
Lines | 84.743% (`461/544`) | :-1: -1.183
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

kush-elastic reviewed (approved):
LGTM!

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/95)
* Commit: 485593580b02c81858034d46cda6e7ddfad7a4a2

### Failed CI Steps
* [:pipeline::arrow_up: Upload Pipeline: .buildkite/pipeline.yml](https://buildkite.com/elastic/integrations/builds/95#0188b9e7-1a60-4f99-a341-73d78edfbfa8)

cc @harnish-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/95","number":95,"commit":"485593580b02c81858034d46cda6e7ddfad7a4a2"}],"number":95}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/169)
* Commit: 2a71c403955b9d411e4024df919599f6f251b6b9

### Failed CI Steps
* [:pipeline::arrow_up: Upload Pipeline: .buildkite/pipeline.yml](https://buildkite.com/elastic/integrations/builds/169#0188c3a3-20b9-4e52-becc-e2f90915a0cf)

### History
* :broken_heart: [Build #95](https://buildkite.com/elastic/integrations/builds/95) failed 485593580b02c81858034d46cda6e7ddfad7a4a2

cc @harnish-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/169","number":169,"commit":"2a71c403955b9d411e4024df919599f6f251b6b9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/95","number":95,"commit":"485593580b02c81858034d46cda6e7ddfad7a4a2"}],"number":169}
buildkite-pr-comment-->

harnish-elastic commented:
/test

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/172)
* Commit: 2a71c403955b9d411e4024df919599f6f251b6b9

### Failed CI Steps
* [:pipeline::arrow_up: Upload Pipeline: .buildkite/pipeline.yml](https://buildkite.com/elastic/integrations/builds/172#0188c3f0-7fa6-426c-811b-8c6f34bf5161)

### History
* :broken_heart: [Build #169](https://buildkite.com/elastic/integrations/builds/169) failed 2a71c403955b9d411e4024df919599f6f251b6b9
* :broken_heart: [Build #95](https://buildkite.com/elastic/integrations/builds/95) failed 485593580b02c81858034d46cda6e7ddfad7a4a2

cc @harnish-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/172","number":172,"commit":"2a71c403955b9d411e4024df919599f6f251b6b9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/169","number":169,"commit":"2a71c403955b9d411e4024df919599f6f251b6b9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/95","number":95,"commit":"485593580b02c81858034d46cda6e7ddfad7a4a2"}],"number":172}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/173)
* Commit: 4365503fe611a927f1c00caf50bc4af5f953081c

### Failed CI Steps
* [:pipeline::arrow_up: Upload Pipeline: .buildkite/pipeline.yml](https://buildkite.com/elastic/integrations/builds/173#0188c3f5-d84b-4e0a-ae28-271a4a0b979c)

### History
* :broken_heart: [Build #172](https://buildkite.com/elastic/integrations/builds/172) failed 2a71c403955b9d411e4024df919599f6f251b6b9
* :broken_heart: [Build #169](https://buildkite.com/elastic/integrations/builds/169) failed 2a71c403955b9d411e4024df919599f6f251b6b9
* :broken_heart: [Build #95](https://buildkite.com/elastic/integrations/builds/95) failed 485593580b02c81858034d46cda6e7ddfad7a4a2

cc @harnish-elastic

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/173","number":173,"commit":"4365503fe611a927f1c00caf50bc4af5f953081c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/172","number":172,"commit":"2a71c403955b9d411e4024df919599f6f251b6b9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/169","number":169,"commit":"2a71c403955b9d411e4024df919599f6f251b6b9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/95","number":95,"commit":"485593580b02c81858034d46cda6e7ddfad7a4a2"}],"number":173}
buildkite-pr-comment-->

ishleenk17 commented on packages/apache_tomcat/data_stream/memory/manifest.yml:
```diff
@@ -0,0 +1,33 @@
+        multi: false
+        required: false
+        show_user: false
+        description: Processors are used to reduce the number of fields in the exported event or to enhance the event with metadata. This executes in the agent before the logs are parsed. See [Processors](https://www.elastic.co/guide/en/beats/filebeat/current/filtering-and-enhancing-data.html) for details.
```
Please link to the Elastic Agent Processors link https://www.elastic.co/guide/en/fleet/current/elastic-agent-processor-configuration.html

harnish-elastic replied:
Updated, thanks!

harnish-elastic commented on packages/apache_tomcat/data_stream/memory/fields/fields.yml:
```diff
@@ -0,0 +1,78 @@
+                  metric_type: gauge
+                - name: time.ms
+                  type: long
+                  description: The amount of time (in milliseconds) that garbage collection took during the collection interval.
```
Updated, thanks!

harnish-elastic commented on packages/apache_tomcat/data_stream/memory/fields/fields.yml:
```diff
@@ -0,0 +1,78 @@
+                  unit: ms
+            - name: valid
+              type: long
+              description: Valid Garbage collection.
```
Updated, thanks!

ishleenk17 commented:
Looks good!
I would wait for Agi/Richa's feedback for tsdb perspective. @agithomas @ritalwar .
@SubhrataK : Any feedback for dashboards ?

agithomas commented on packages/apache_tomcat/data_stream/memory/fields/ecs.yml:
```diff
@@ -0,0 +1,33 @@
+  name: cloud.instance.id
+  dimension: true
+- external: ecs
+  name: cloud.project.id
```
Please refer [here](https://github.com/elastic/integrations/issues/5193#issuecomment-1536090359) for the latest common dimension fields. Please update the mappings for all datastreams.

harnish-elastic replied:
Updated, thanks!

harnish-elastic commented on packages/apache_tomcat/data_stream/memory/fields/fields.yml:
```diff
@@ -0,0 +1,78 @@
+                - name: count
+                  type: long
+                  description: The cumulative number of invoked garbage collections since the start of the server.
+                  metric_type: gauge
```
Updated, thanks!

agithomas reviewed (approved):
LGTM considering only TSDB.

ishleenk17 reviewed (approved):
Looks good!

harnish-elastic commented:
As discussed with @SubhrataK regarding the memory dashboard, re-arrange the order of visualizations, updated metrics chart visualizations and updated color for Garbage collection over time chart for better readability.
![memory-dashboard](https://github.com/elastic/integrations/assets/118714680/4b2fb995-5569-470f-8e53-a1f0bc0df55b)

SubhrataK reviewed (approved):
Looks good.

elasticmachine commented:
Package apache_tomcat - 0.9.0 containing this change is available at https://epr.elastic.co/search?package=apache_tomcat
